### PR TITLE
[CausalLM] Implement and Integrate Ernie 4.5 MoE Model

### PR DESCRIPTION
**Pull request by DonghakPark**
## Dependency of the PR
- None
## Commits to be reviewed in this PR

<details><summary>[CausalLM] Implement ERNIE's GLM Style RoPE</summary><br />

    [CausalLM] Implement ERNIE's GLM Style RoPE
    
    Implement GLM Sytle RoPE at MHA CORE
    
    **Self evaluation:**
    1. Build test:   [X]Passed [ ]Failed [ ]Skipped
    2. Run test:     [X]Passed [ ]Failed [ ]Skipped
    
    Signed-off-by: Donghak PARK <donghak.park@samsung.com>

</details>

<details><summary>[CausalLM] Add ernie to main & meson build</summary><br />

    [CausalLM] Add ernie to main & meson build
    
    Add ernie model & Layer to main, meson build
    
    **Self evaluation:**
    1. Build test:   [X]Passed [ ]Failed [ ]Skipped
    2. Run test:     [X]Passed [ ]Failed [ ]Skipped
    
    Signed-off-by: Donghak PARK <donghak.park@samsung.com

</details>

<details><summary> [CausalLM] Implement Erine MoE Layer</summary><br />

    [CausalLM] Implement Erine MoE Layer
    
    Implement Ernie MoE Layer
    - Shared Expert accum
    - static bias add
    
    **Self evaluation:**
    1. Build test:   [X]Passed [ ]Failed [ ]Skipped
    2. Run test:     [X]Passed [ ]Failed [ ]Skipped
    
    Signed-off-by: Donghak PARK <donghak.park@samsung.com>

</details>

<details><summary>[CausalLM] Add causallm common properties <num_shared experts, moe_norm_min></summary><br />

    [CausalLM] Add causallm common properties <num_shared experts, moe_norm_min>
    
    add causallm common properties
    - num_shared_experts
    - moe_norm_min
    
    **Self evaluation:**
    1. Build test:   [X]Passed [ ]Failed [ ]Skipped
    2. Run test:     [X]Passed [ ]Failed [ ]Skipped
    
    Signed-off-by: Donghak PARK <donghak.park@samsung.com>

</details>

<details><summary>[Application][CausalLM] Implement Ernie 4.5 MoE Model</summary><br />

    [Application][CausalLM] Implement Ernie 4.5 MoE Model
    
    Implemnet Ernie 4.5 MoE Model
    - ernie's first layer is dense
    - ernie has shared expert at each MoE Layer
    
    **Self evaluation:**
    1. Build test:   [X]Passed [ ]Failed [ ]Skipped
    2. Run test:     [X]Passed [ ]Failed [ ]Skipped
    
    Signed-off-by: Donghak PARK <donghak.park@samsung.com>

</details>

## Summary

This PR implements and integrates the Ernie 4.5 MoE (Mixture of Experts) model into the CausalLM application. 
The changes include the implementation of the model structure, the MoE layer, GLM-style RoPE, and build system integration.

### Erine 4.5 Models Key Difference btw Qwen

**1. Qwen apply RMSNorm before RoPE to Q, K but Ernie don't**

**2. Ernie Apply GLM Style RoPE** 

```.py

def rotate_half(x):
    """Rotates half the hidden dims of the input."""
    x1 = x[..., 0::2]
    x2 = x[..., 1::2]
    return torch.stack((-x2, x1), dim=-1).flatten(-2)


def apply_rotary_pos_emb(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
    """Applies Rotary Position Embedding to the query and key tensors.

    Args:
        q (`torch.Tensor`): The query tensor.
        k (`torch.Tensor`): The key tensor.
        cos (`torch.Tensor`): The cosine part of the rotary embedding.
        sin (`torch.Tensor`): The sine part of the rotary embedding.
        position_ids (`torch.Tensor`, *optional*):
            Deprecated and unused.
        unsqueeze_dim (`int`, *optional*, defaults to 1):
            The 'unsqueeze_dim' argument specifies the dimension along which to unsqueeze cos[position_ids] and
            sin[position_ids] so that they can be properly broadcasted to the dimensions of q and k. For example, note
            that cos[position_ids] and sin[position_ids] have the shape [batch_size, seq_len, head_dim]. Then, if q and
            k have the shape [batch_size, heads, seq_len, head_dim], then setting unsqueeze_dim=1 makes
            cos[position_ids] and sin[position_ids] broadcastable to the shapes of q and k. Similarly, if q and k have
            the shape [batch_size, seq_len, heads, head_dim], then set unsqueeze_dim=2.
    Returns:
        `tuple(torch.Tensor)` comprising of the query and key tensors rotated using the Rotary Position Embedding.
    """
    # glm rope style (with full dim) and full precision
    original_dtype = q.dtype

    cos = cos.unsqueeze(unsqueeze_dim)
    sin = sin.unsqueeze(unsqueeze_dim)

    # Interleave them instead of usual shape
    cos = cos[..., : cos.shape[-1] // 2].repeat_interleave(2, dim=-1)
    sin = sin[..., : sin.shape[-1] // 2].repeat_interleave(2, dim=-1)

    q_embed = (q.float() * cos) + (rotate_half(q).float() * sin)
    k_embed = (k.float() * cos) + (rotate_half(k).float() * sin)

    return q_embed.to(original_dtype), k_embed.to(original_dtype)
```

**3. Ernie use 2 Shared Expert + 6 TopK Expert (in 21B-A3B Model)**

**4. Ernie Has ```e_score_correction_bias``` & add this bias after softmax of router**

```.py
router_logits = F.linear(hidden_states.float(), self.weight)
router_logits = F.softmax(router_logits, dim=1, dtype=torch.float)
router_top_value, router_indices = torch.topk(self.moe_statics(router_logits), self.top_k, dim=-1)
```

Signed-off-by: Donghak PARK <donghak.park@samsung.com>

### CODE REVIEW

File: Applications/CausalLM/layers/ernie_moe_layer.cpp [hunk 1]
```diff
@@ -0,0 +1,520 @@
...
+        std::lock_guard<std::mutex> lock(cache_mutex);
+        target_idx = loaded_expert_deque.front();
+        loaded_expert_deque.pop_front();
+        iteration_map.erase(target_idx);
+        need_load[target_idx] = true;
+      }
+      context.getWeight(expert_gate_proj_indices[target_idx]).deactivate();
+      context.getWeight(expert_up_proj_indices[target_idx]).deactivate();
+      context.getWeight(expert_down_proj_indices[target_idx]).deactivate();
+    }
```
Comment (skykongkong8) [inline Applications/CausalLM/layers/ernie_moe_layer.cpp:477]: The while loop condition accesses loaded_expert_deque.size() without a lock, while other threads (or even the current thread in a different context) might be modifying it. This is a data race. Furthermore, checking the size outside the lock and then popping inside creates a race condition where the deque could become empty between the check and the front() call.

How about fixing lock_guard like:
```C++
#pragma omp parallel
while (true) {
  int target_idx;
  {
    std::lock_guard<std::mutex> lock(cache_mutex);
    if (loaded_expert_deque.size() <= 16) break; 
    target_idx = loaded_expert_deque.front();
    loaded_expert_deque.pop_front();
    // ... update maps/flags
  }
  // ... deactivate weights (outside lock)
}
```

Comment (DonghakPark) [reply Applications/CausalLM/layers/ernie_moe_layer.cpp:477]: oh!!! thank you, i fixed not only this Model, also all other models too.

Comment (skykongkong8) [reply Applications/CausalLM/layers/ernie_moe_layer.cpp:477]: That's awesome! 😊 Glad to hear you've applied the fix across all models—definitely a proactive move. If you'd like me to review any other parts of the code or help with testing, just give a shout. Great teamwork! 💪

File: Applications/CausalLM/layers/gpt_oss_moe_layer_cached.cpp [hunk 2]
```diff
@@ -378,10 +378,13 @@ void CachedSlimGptOssMoELayer::incremental_forwarding(
 
 // Evict experts
 #pragma omp parallel
-    while (loaded_expert_deque.size() > 16) {
+    while (true) {
       int target_idx;
       {
         std::lock_guard<std::mutex> lock(cache_mutex);
+        if (loaded_expert_deque.size() > 16)
```
Comment (myungjoo) [inline Applications/CausalLM/layers/gpt_oss_moe_layer_cached.cpp:385]: 1. ```>``` --> ```<``` ? or ```<=```?
2. need a comment or properly-named constant for 16. (what does it mean)

File: Applications/CausalLM/layers/qwen_moe_layer_cached.cpp [hunk 3]
```diff
@@ -438,10 +438,12 @@ void CachedSlimMoELayer::incremental_forwarding(
 
 // Evict experts
 #pragma omp parallel
-    while (loaded_expert_deque.size() > 32) {
+    while (true) {
       int target_idx;
       {
         std::lock_guard<std::mutex> lock(cache_mutex);
+        if (loaded_expert_deque.size() > 16)
```
Comment (myungjoo) [inline Applications/CausalLM/layers/qwen_moe_layer_cached.cpp:445]: 1. BUG?: 16? 32? >? <? <=?
2. need a proper constant name or a comment explaining wth is 16/32.

File: Applications/CausalLM/layers/ernie_moe_layer.cpp [hunk 4]
```diff
@@ -0,0 +1,524 @@
...
+      float sum_prob = 0.0f;
+      for (int k = 0; k < static_cast<int>(topk); ++k) {
+        unsigned expert_idx = indices_data[i * topk + k];
+        float weight = router_logits.getValue<float>(i, 0, 0, expert_idx);
+        sum_prob += weight;
+      }
+
+      for (int k = 0; k < static_cast<int>(topk); ++k) {
+        unsigned expert_idx = indices_data[i * topk + k];
+        float weight = router_logits.getValue<float>(i, 0, 0, expert_idx);
```
Comment (myungjoo) [inline Applications/CausalLM/layers/ernie_moe_layer.cpp:398]: Is there any way to reuse the values from the previous loop?

File: Applications/CausalLM/layers/ernie_moe_layer.cpp [hunk 5]
```diff
@@ -0,0 +1,524 @@
...
+
+    const uint32_t *indices_data = topk_indices.getData<uint32_t>();
+    std::vector<std::vector<std::pair<unsigned, float>>> expert_assignments(
+      num_experts);
+    // Set expert mask
+    for (int i = 0; i < static_cast<int>(total_tokens); ++i) {
+      float sum_prob = 0.0f;
+      for (int k = 0; k < static_cast<int>(topk); ++k) {
+        unsigned expert_idx = indices_data[i * topk + k];
+        float weight = router_logits.getValue<float>(i, 0, 0, expert_idx);
```
Comment (myungjoo) [inline Applications/CausalLM/layers/ernie_moe_layer.cpp:392]: Is it ok to keep use router_logits? instead of bias_router_logits? or need ``` rope_scaling_type == "glm"``` somewhere?

File: Applications/CausalLM/layers/mha_core.h [hunk 6]
```diff
@@ -422,6 +422,27 @@ WIN_EXPORT class MHACoreLayer : public nntrainer::LayerImpl {
...
+
+  /**
+   * @brief     apply rotary embedding for ERNIE
+   * @param[in] in input tensor
+   * @param[out] out output tensor
+   * @param[in] dim hidden dim size
+   * @param[in] from sequence order
+   * @param[in] convert_only - conversion only
+   */
+  void apply_rotary_emb_tensor_ernie(nntrainer::Tensor &in,
```
Comment (myungjoo) [inline Applications/CausalLM/layers/mha_core.h:442]: looks like this function is unused? non GLM-style RoPE?

File: Applications/CausalLM/layers/ernie_moe_layer.cpp [hunk 7]
```diff
@@ -0,0 +1,524 @@
...
+          context.getWeight(expert_down_proj_indices[expert_idx]), hidden_size);
+
+      } else {
+
+        compute_expert_forward(
+          input, expert_outputs[expert_idx], assignments,
+          context.getWeight(expert_gate_proj_indices[expert_idx]),
+          context.getWeight(expert_up_proj_indices[expert_idx]),
+          context.getWeight(expert_down_proj_indices[expert_idx]), hidden_size);
+      }
```
Comment (myungjoo) [inline Applications/CausalLM/layers/ernie_moe_layer.cpp:456]: Check the above logic (it's a FIFO!!). What happens if one or a few hot experts are ALWAYS hit and there are other experts used sporadically for more than 16 times? Aren't the hot experts evicted even if they are always used for 16 times?

Even when need_load is false, you may need to re-push the hit expert. (to keep it alive for the next 16 accesses).
